doc: add explanation and guide for "bases" (#597)
canonical/rockcraft: docs/explanation/bases.rst

.. _bases_explanation:

Bases
=====

One mandatory field for every rock project is the
:ref:`base <rockcraft_yaml_base>`. 

This field defines the baseline system upon which the rock's contents shall be
layered on. Only Ubuntu and ``bare``
:ref:`bases <rockcraft_yaml_base>` are supported.

Ubuntu bases
------------

When setting an Ubuntu
:ref:`base <rockcraft_yaml_base>` for a rock, Rockcraft will pull the
corresponding Ubuntu base container image from a container registry. This is
exactly the same container one would get from places like
`Docker Hub <docker_hub_>`_ or `Amazon ECR <ecr_>`_.

> The resulting rock will therefore also contain and software and utilities that

Reviewer: ```suggestion
The resulting rock will therefore also contain any software and utilities that
```

are present in the Ubuntu base container images.

Setting an Ubuntu :ref:`base <rockcraft_yaml_base>` for a rock is
especially useful when the goal is to build a rock that can be used for
general-purpose environments such as a development working space.

Bare bases
----------

> As the name says, a ``bare`` :ref:`base <rockcraft_yaml_base>` indicates

Reviewer: ```suggestion
As the name suggests, a ``bare`` :ref:`base <rockcraft_yaml_base>` indicates
```

that the rock shall have no

> baseline system. This definition is similar to the "`scratch`_" Docker image,
> with the exception that a rock **can never be completely empty**, since it must
> always have :ref:`Pebble <pebble_explanation_page>` as its entrypoint, and
> include their own :ref:`metadata <what-sets-rocks-apart>` inside.

Reviewer 2: I think you are referring to the rock with _their_, there is too much indirection here as _their_ could be referring to Pebble for the unsuspecting victim and only figure it out by reading the metadata link.

```suggestion
baseline system. This definition is similar to the "`scratch`_" Docker image,
with the exception that every rocks itself always includes a minimum baseline
such as  :ref:`Pebble <pebble_explanation_page>`  and additional metadata:
:ref:`<what-sets-rocks-apart>`.
```
I would not alias the link for what sets the rock apart and let the original title render (I might have not gotten the syntax right)

Author: Sure. Applied the suggestion with some re-wording.

When and how to use a ``bare`` base
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The ``bare`` :ref:`base <rockcraft_yaml_base>` is extremely useful when
the goal is to build a lean **chiselled rock**. Such rocks are typically meant
for production environments, and as such, the combination of the ``bare``
:ref:`base <rockcraft_yaml_base>`
with :ref:`Chisel <chisel_explanation>` slices will result in a minimalistic
container image that has a reduced attack surface while retaining its

> functionality.

Reviewer 2: perhaps words on why these rocks are meant for production environments, or perhaps preferred is the better word.

I would go along the lines of reducing the surface area and the security vectors

Author: Sure. Re-worded it
